File: src/test/java/ez/spring/vertx/web/TestWebHandler.java
package ez.spring.vertx.web;

import org.springframework.stereotype.Component;

import ez.spring.vertx.web.handler.WebHandler;
import ez.spring.vertx.web.handler.request.RequestReader;
import ez.spring.vertx.web.handler.response.ResponseWriter;
import io.vertx.core.Future;

@Component
public class TestWebHandler extends WebHandler<TestWebHandler.Data, TestWebHandler.Data> {
  @Override
  public Future<Data> exec(Data data) {
    Data r = new Data();
    r.setValue(data.getValue());
    return Future.succeededFuture(r);
  }

  @Override
  public RequestReader<Data> getRequestReader() {
    return RequestReader.jsonBody(Data.class);
  }

  @Override
  public ResponseWriter<Data> getResponseWriter() {
    return ResponseWriter.json();
  }

  @SuppressWarnings("WeakerAccess")
  static class Data {
    private String value = "";

    public String getValue() {
      return value;
    }

    public void setValue(String value) {
      this.value = value;
    }
  }
}

File: src/test/java/ez/spring/vertx/web/TestWebHandler.kt
package ez.spring.vertx.web

import ez.spring.vertx.web.handler.AsyncWebHandler
import ez.spring.vertx.web.handler.request.RequestReader
import ez.spring.vertx.web.handler.response.JsonResponseWriter
import ez.spring.vertx.web.handler.response.ResponseWriter
import kotlinx.coroutines.delay
import org.slf4j.LoggerFactory
import java.util.concurrent.ThreadLocalRandom

class TestWebHandler : AsyncWebHandler<TestWebHandler.Body, TestWebHandler.Body>() {
    private val log = LoggerFactory.getLogger(javaClass)
    override suspend fun execAsync(request: Body?): Body? {
        log.info("res1 {}", request?.b)
        delay(200 + (ThreadLocalRandom.current().nextDouble() * 400).toLong())
        log.info("res2 {}", request?.b)
        return request
    }

    override val requestReader: RequestReader<Body>?
        get() = RequestReader.jsonBody(Body::class.java)

    override val responseWriter: JsonResponseWriter<Body>?
        get() = ResponseWriter.json()

    class Body {
        var a = 0
        var b: String? = null
    }
}